src/Common/QueryProfiler.cpp

```diff
@@ -85,6 +85,92 @@ namespace ErrorCodes
     extern const int NOT_IMPLEMENTED;
 }
 
+#if USE_UNWIND
+Timer::Timer():log(&Poco::Logger::get("Timer")){}
+
+void Timer::createIfNecessary(UInt64 thread_id, int clock_type, int pause_signal)
+{
+    if (!timer_id)
+    {
+        struct sigevent sev {};
+        sev.sigev_notify = SIGEV_THREAD_ID;
+        sev.sigev_signo = pause_signal;
+
+#if defined(OS_FREEBSD)
+        sev._sigev_un._threadid = static_cast<pid_t>(thread_id);
+#elif defined(USE_MUSL)
+        sev.sigev_notify_thread_id = static_cast<pid_t>(thread_id);
+#else
+        sev._sigev_un._tid = static_cast<pid_t>(thread_id);
+#endif
+        timer_t local_timer_id;
+        if (timer_create(clock_type, &sev, &local_timer_id))
+        {
+            /// In Google Cloud Run, the function "timer_create" is implemented incorrectly as of 2020-01-25.
+            /// https://mybranch.dev/posts/clickhouse-on-cloud-run/
+            if (errno == 0)
+                throw Exception(ErrorCodes::CANNOT_CREATE_TIMER, "Failed to create thread timer. The function "
+                                "'timer_create' returned non-zero but didn't set errno. This is bug in your OS.");
+
+            throwFromErrno("Failed to create thread timer", ErrorCodes::CANNOT_CREATE_TIMER);
+        }
+        timer_id.emplace(local_timer_id);
+    }
+}
+
+void Timer::set(UInt32 period)
+{
+    /// Too high frequency can introduce infinite busy loop of signal handlers. We will limit maximum frequency (with 1000 signals per second).
+    if (period < 1000000)
+        period = 1000000;
+    /// Randomize offset as uniform random value from 0 to period - 1.
+    /// It will allow to sample short queries even if timer period is large.
+    /// (For example, with period of 1 second, query with 50 ms duration will be sampled with 1 / 20 probability).
+    /// It also helps to avoid interference (moire).
+    UInt32 period_rand = std::uniform_int_distribution<UInt32>(0, period)(thread_local_rng);
+
+    struct timespec interval{.tv_sec = period / TIMER_PRECISION, .tv_nsec = period % TIMER_PRECISION};
+    struct timespec offset{.tv_sec = period_rand / TIMER_PRECISION, .tv_nsec = period_rand % TIMER_PRECISION};
+
+    struct itimerspec timer_spec = {.it_interval = interval, .it_value = offset};
+    if (timer_settime(*timer_id, 0, &timer_spec, nullptr))
+        throwFromErrno("Failed to set thread timer period", ErrorCodes::CANNOT_SET_TIMER_PERIOD);
+}
+
+void Timer::stop()
+{
+    if (timer_id)
+    {
+        struct timespec stop_timer{.tv_sec = 0, .tv_nsec = 0};
+        struct itimerspec timer_spec = {.it_interval = stop_timer, .it_value = stop_timer};
+        int err = timer_settime(*timer_id, 0, &timer_spec, nullptr);
+        if (err)
+            LOG_ERROR(log, "Failed to stop query profiler timer {}", errnoToString());
+        chassert(!err && "Failed to stop query profiler timer");
```

Comment:
```suggestion
        if (err)
            throwFromErrno("Failed to stop query profiler timer", ErrorCodes::LOGICAL_ERROR);
```

You may think about `LOGICAL_ERROR` as about `chassert`